src/NHapi.Base/Parser/PipeParser.cs

```diff
@@ -24,885 +24,960 @@
+                    }
+
+                    throw new HL7Exception(buf.ToString(), ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+                }
+            }
+            catch (IndexOutOfRangeException e)
+            {
+                throw new HL7Exception("Can't find message structure (MSH-9-3): " + e.Message, ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+            }
+
+            return new MessageStructure(messageStructure, explicityDefined);
+        }
+
+        /// <summary>
+        /// A struct for holding a message class string and a boolean indicating whether it
+        /// was defined explicitly.
+        /// </summary>
+        private class MessageStructure
+        {
+            public MessageStructure(string theMessageStructure, bool isExplicitlyDefined)
+            {
+                Structure = theMessageStructure;
+                ExplicitlyDefined = isExplicitlyDefined;
+            }
+
+            public string Structure { get; }
+
+            public bool ExplicitlyDefined { get; }
+        }
+
+        private class AnonymousClassPredicate : FilterIterator.IPredicate
+        {
+            private PipeParser enclosingInstance;
+
+            public AnonymousClassPredicate(PipeParser enclosingInstance)
+            {
+                InitBlock(enclosingInstance);
+            }
+
+            public PipeParser Enclosing_Instance
+            {
+                get { return enclosingInstance; }
+            }
+
+            public virtual bool evaluate(object obj)
+            {
+                return Evaluate(obj);
+            }
+
+            public virtual bool Evaluate(object obj)
+            {
+                return typeof(ISegment).IsAssignableFrom(obj.GetType());
+            }
+
+            private void InitBlock(PipeParser enclosingInstance)
+            {
+                this.enclosingInstance = enclosingInstance;
+            }
+        }
+
+        private class AnonymousClassPredicate1 : FilterIterator.IPredicate
+        {
+            private string name;
+            private PipeParser enclosingInstance;
+
+            public AnonymousClassPredicate1(string name, PipeParser enclosingInstance)
+            {
+                InitBlock(name, enclosingInstance);
+            }
+
+            public PipeParser Enclosing_Instance
+            {
+                get { return enclosingInstance; }
+            }
+
+            /// <inheritdoc />
+            public virtual bool evaluate(object obj)
+            {
+                return Evaluate(obj);
+            }
+
+            /// <inheritdoc />
+            public virtual bool Evaluate(object obj)
+            {
+                var structureName = ((IStructure)obj).GetStructureName();
+                Log.Debug($"PipeParser iterating message in direction {name} at {structureName}");
+
+                return Regex.IsMatch(structureName, $"{name}\\d*");
+            }
+
+            private void InitBlock(string name, PipeParser enclosingInstance)

```

Comment:
same with this one, can we kill this method and move assignment to constructor?

Reply:
done